### PR TITLE
tracking: adding a mosse tracker

### PR DESCRIPTION
**Pull request by berak**
this pr adds a MOSSE tracker to the tracking module, quite similar to the python sample in opencv main.

### CODE REVIEW

**Comment by berak** [comment]
```
benchmark -gt=david.gt.txt -video=david.webm -start=300 MOSSE

==========
mosse
       Overlap > 0                   100%                 469
       Overlap > 0.5             57.1429%                 268
           Precision             57.1429%
           Recall                57.1429%
           f-measure             57.1429%
                 AUC            0.509765
         Performance             2.23582 ms/frame             447.262 fps

```

(tests were added, too. since it (visually) behaves similar to the MIL one, i copied the resp. values from there, maybe someone can help finding better ones.)

**Comment by sovrasov** [comment]
@berak to find the threshold values just run the tests locally and see the actual values in error messages:
```
Tracking/DistanceAndOverlap.Scaled_Data_MOSSE/0 ["david"] : 
/build/precommit-contrib_linux64/opencv_contrib/modules/tracking/test/test_trackers.cpp:312: Failure
Failed
Incorrect overlap: curr = 0.4690278172492981, min = 0.5

Google Test trace:
/build/precommit-contrib_linux64/opencv_contrib/modules/tracking/test/test_trackers.cpp:407: A
Tracking/DistanceAndOverlap.Scaled_Data_MOSSE/1 ["dudek"] : 
/build/precommit-contrib_linux64/opencv_contrib/modules/tracking/test/test_trackers.cpp:306: Failure
Failed
Incorrect distance: curr = 44.406688690185547, max = 25

Google Test trace:
/build/precommit-contrib_linux64/opencv_contrib/modules/tracking/test/test_trackers.cpp:407: A
```
For overlap I recommend to use actual - 0.05
for distance: actual + 10
Sometimes they differ on different platforms, then use the buildbot for that.

File: modules/tracking/src/mosseTracker.cpp [hunk 1]
```diff
@@ -0,0 +1,291 @@
+/*///////////////////////////////////////////////////////////////////////////////////////
```
Comment (sovrasov) [inline modules/tracking/src/mosseTracker.cpp]: Please use short license header for the new files:
```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
/*
```

**Comment by sovrasov** [comment]
Also MOSSE should be added in [this](https://github.com/opencv/opencv_contrib/blob/6ffdd0f4ba9ab39cba200d487efac3f73ac44937/modules/tracking/samples/samples_utility.hpp) function to be able to use the new tracker from the `tracker` sample.

**Comment by berak** [comment]
@sovrasov , thanks for kindly looking at it !

>  run the tests locally 
had some problems, obtaining the data, initially, but that's solved now. let's see.

> use short license header
sure.

> Also MOSSE should be added in this function
i think, i have it [covered already](https://github.com/opencv/opencv_contrib/pull/1423/files#diff-1519ded1b562802bd2ee88c326058ac7R23)

File: modules/tracking/test/test_trackers.cpp [hunk 2]
```diff
@@ -523,7 +536,7 @@ TEST_P(DistanceAndOverlap, Scaled_Data_KCF)
 
 TEST_P(DistanceAndOverlap, DISABLED_Scaled_Data_TLD)
 {
-  TrackerTest test( TrackerTLD::create(), dataset, 120, .45f, Scale_1_1);
+  TrackerTest test( TrackerTLD::create(), dataset, 30, .7f, Scale_1_1);
```
Comment (sovrasov) [inline modules/tracking/test/test_trackers.cpp]: Please revert this change. It's not related to mosse and will conflict withe the existed pull request related to TLD.

Comment (berak) [reply modules/tracking/test/test_trackers.cpp]: oops, ofc. !

**Comment by sovrasov** [comment]
I've obtained two different results on the dudek video:
```
MOSSE
       Overlap > 0                20.542%                 235
       Overlap > 0.5             17.7448%                 203
           Precision             17.7914%
           Recall                18.9366%
           f-measure             18.3461%
                 AUC            0.874668
         Performance             1.25915 ms/frame             794.188 fps

```
and
```
MOSSE
       Overlap > 0               93.7063%                1072
       Overlap > 0.5             66.6958%                 763
           Precision             66.6958%
           Recall                71.1754%
           f-measure             68.8628%
                 AUC            0.474895
         Performance             1.23888 ms/frame             807.184 fps
```
In first case the tracker just failed. Are there any ideas why results are non-deterministic?

File: modules/tracking/test/test_trackers.cpp [hunk 3]
```diff
@@ -534,6 +547,13 @@ TEST_P(DistanceAndOverlap, DISABLED_GOTURN)
   test.run();
 }
 
+TEST_P(DistanceAndOverlap, Scaled_Data_MOSSE)
+{
+  TrackerTest test( TrackerMOSSE::create(), dataset, 22, -0.05f, Scale_1_1, 1);
```
Comment (sovrasov) [inline modules/tracking/test/test_trackers.cpp]: Why do you have a negative value here? If the tracker failed, 0 is enough. I have good results on Ubuntu 16.04 in this test (min overlap is 0.69).

Comment (berak) [reply modules/tracking/test/test_trackers.cpp]: oh, i simply must have misread your comment about the values above.

i'll try again ;)

**Comment by berak** [comment]
indeed, [randomWarp, here](https://github.com/opencv/opencv_contrib/pull/1423/files#diff-9df63a86c375d7d9982b9379892e61eaR111)

should i use a fixed seed instead ?

**Comment by sovrasov** [comment]
Yes, because the tests can accidentally fail on buldbot (in case of "bad" seed parameter). This is no good.

**Comment by berak** [comment]
ok, got it. 

dudek is unfair, gt is off a lot ;)

**Comment by sovrasov** [comment]
:+1:
